### DISCUSSION TITLE
Wrong bounding box after creating a selection.

### DISCUSSION THREAD

**Discussion by jeroenrosendaal** [fabricjs/fabric.js#10215] [Q&A] [answered]
When selecting all objects inside the canvas and creating an `ActiveSelection` of it the bounding box becomes wrong. Does somebody know how to fix this? If read about `.setCoords()` but haven't succeeded with this.

[https://jsfiddle.net/vhbx6prk/3/](https://jsfiddle.net/vhbx6prk/3/)

![Screenshot 2024-10-16 at 10 34 12](https://github.com/user-attachments/assets/b8b552d9-c4e4-43ec-b608-7ed6386a9482)

```js
const canvas = new fabric.Canvas("canvas")
canvas.setWidth(document.body.clientWidth)
canvas.setHeight(document.body.clientHeight)

canvas.on('after:render', () => {
	canvas.contextContainer.strokeStyle = '#555';
  canvas.forEachObject((obj) => {
    var bound = obj.getBoundingRect();
    canvas.contextContainer.strokeRect(
      bound.left + 0.5,
      bound.top + 0.5,
      bound.width,
      bound.height
    );
  });
});

const word1 = new fabric.Text("Example 1", {
  left: 100,
  top: 100,
  width: 50,
  height: 56,
})

const word2 = new fabric.Text("Example 1", {
  left: 200,
  top: 200,
  width: 50,
  height: 56,
  fill: "red",
})

canvas.add(word1)
canvas.add(word2)

const selection = new fabric.ActiveSelection(canvas.getObjects(), {
  canvas: canvas,
});


```

**Comment by asturur** [comment]
In fabric 5 the getBounding box is relative to the parent.
You need to use the activeSelection transform to transform the bounding box.

I suggest you to see how the method is built in fabric 6 and just copy the code in your project if you need it in that way. If you are at the start of your project use fabric 6.

**Reply by jeroenrosendaal** [reply]
Thank you for your fast reply. What method your are revering to? Do you mean the `ActiveSelection`?

**Reply by asturur** [reply]
no i mean `getBoundingRect` is based on Object.getCoords(). Object.getCoords() returns you coords that are relative to the container ( in this case activeSelection ).
In fabric6 this change and should work out of the box.
You can read the change code of `getCoords` in fabric 6.x and apply the code into your project.

**Reply by jeroenrosendaal** [reply]
I see, thanks. You would recommend something like this to update the method? I can't currently update to version 6.

```js
fabric.Object.prototype.getCoords = (function() {
  return function(...args) {
  	const { tl, tr, br, bl } = this.aCoords || (this.aCoords = this.calcACoords());
    const coords = [tl, tr, br, bl];
    if (this.group) {
      const t = this.group.calcTransformMatrix();
      return coords.map((p) => transformPoint(p, t));
    }
    return coords;
  };
})(fabric.Canvas.prototype.getCoords);
```

**Reply by jeroenrosendaal** [reply]
This will force me to add a lot of methods that doesn't exists right now.

**Reply by asturur** [reply]
You can just write your own getBoundingBox function it doesn't have to be on the prototype.
function myBoundingBox(object) {
  // get corrds
  // transform them
  // calculate the bounding box from the points
}

**Reply by jeroenrosendaal** [reply]
@asturur I've composed the method below. It's working outside a custom object. If I use it inside a custom object that renders a canvas (like an image) there is an offset. Are you able to see what I'm missing? I have copied the code from the source code of fabric.js 6 and added canvas.getZoom() at the and. I'm thinking of a kind of scaling that I'm missing.

![Screenshot 2024-10-21 at 11 22 32](https://github.com/user-attachments/assets/12dcd825-d0d6-4d34-a07c-3f5615ac6e08)

``` js
function getBoundingRect(obj) {
  const { tl, tr, br, bl } = obj.aCoords || (obj.aCoords = obj.calcACoords());
  let coords = [tl, tr, br, bl];
  if (obj.group) {
    const t = obj.group.calcTransformMatrix();
    coords = coords.map((p) => {
      const ignoreOffset = false;
      return new fabric.Point(
        t[0] * p.x + t[2] * p.y + (ignoreOffset ? 0 : t[4]),
        t[1] * p.x + t[3] * p.y + (ignoreOffset ? 0 : t[5]),
      );
    });
  }

  let left = 0,
      top = 0,
      width = 0,
      height = 0;

  for (let i = 0, len = coords.length; i < len; i++) {
    const { x, y } = coords[i];
    if (x > width || !i) width = x;
    if (x < left || !i) left = x;
    if (y > height || !i) height = y;
    if (y < top || !i) top = y;
  }

  return {
    left: left * canvas.getZoom(),
    top: top * canvas.getZoom(),
    width: (width - left) * canvas.getZoom(),
    height: (height - top) * canvas.getZoom(),
  };
}
```

So the hierarchy is as follows:
Fabric canvas
-> custom fabric Object
    -> new Fabric canvas
        -> render some text with bounding box.

**Reply by jeroenrosendaal** [reply]
@asturur [https://jsfiddle.net/6anhw817/1/](https://jsfiddle.net/6anhw817/1/)

**Reply by jeroenrosendaal** [reply]
This fill fix it. But I'm not sure why. Can you explain?

``` js
var mCanvas = this.tmpCanvas.viewportTransform;
var mObject = obj.calcTransformMatrix();
var mTotal = fabric.util.multiplyTransformMatrices(mCanvas, mObject);
const matrix = fabric.util.qrDecompose(mTotal);

 return {
    left: matrix.translateX - (obj.width / 2) * matrix.scaleX,
    top: matrix.translateY - (obj.height / 2) * matrix.scaleY, 
    width: (width - left) * this.tmpCanvas.getZoom(),
    height: (height - top) * this.tmpCanvas.getZoom(),
};
```

**Reply by jeroenrosendaal** [reply]
And this will not take rotation into account. Can you help with that? Can't seem to figure it out.

![Screenshot 2024-10-22 at 09 51 26](https://github.com/user-attachments/assets/6ccdb1fb-a1c1-445c-89ff-fc24b6eeb095)

**Reply by asturur** [reply]
i wonder why you can't use the normal render control box function. seems to me you want the control box.

**Reply by asturur** [reply, marked as answer]
not sure this is what you are  looking for:
https://jsfiddle.net/asturur/2rmuqa5n/3/
